File: support_tools/cpp/upd_commander_checker/tests/single_file_scan_tests.cpp
#include <cassert>
#include <filesystem>
#include <fstream>
#include <string>
#include <vector>

#include "scanner.hpp"

namespace {

void write_file(const std::filesystem::path& path, const std::string& content) {
    std::filesystem::create_directories(path.parent_path());
    std::ofstream file(path);
    file << content;
}

std::string include_file(const std::filesystem::path& path) {
    return "#include \"" + path.generic_string() + "\"\n";
}

bool has_code(const std::vector<upd_checker::Finding>& findings, const std::string& code) {
    for (const auto& finding : findings) {
        if (finding.code == code) {
            return true;
        }
    }
    return false;
}

void test_single_file_preserves_ui_layer() {
    const auto root = std::filesystem::temp_directory_path() / "upd_cpp_single_ui";
    std::filesystem::remove_all(root);
    const auto dependency = root / "applications" / "main" / "data" / "storage.hpp";
    const auto target = root / "applications" / "main" / "ui" / "screen.cpp";
    write_file(dependency, "#pragma once\n");
    write_file(target, include_file(dependency));

    assert(has_code(upd_checker::scan_path(target.string(), {}), "UPD101"));
    std::filesystem::remove_all(root);
}

void test_single_file_preserves_application_boundary() {
    const auto root = std::filesystem::temp_directory_path() / "upd_cpp_single_app";
    std::filesystem::remove_all(root);
    const auto dependency =
        root / "applications" / "settings" / "process" / "settings_processing.hpp";
    const auto target = root / "applications" / "main" / "process" / "main_commander.cpp";
    write_file(dependency, "#pragma once\n");
    write_file(target, include_file(dependency));

    assert(has_code(upd_checker::scan_path(target.string(), {}), "UPD102"));
    std::filesystem::remove_all(root);
}

}  // namespace

int main() {
    test_single_file_preserves_ui_layer();
    test_single_file_preserves_application_boundary();
    return 0;
}

File: support_tools/cs/upd_commander_checker_tests/SingleFileScanTests.cs
using UpdCommanderChecker;

namespace UpdCommanderChecker.Tests;

public sealed class SingleFileScanTests
{
    [Fact]
    public void SingleFileScanPreservesUiLayer()
    {
        using var project = new TempProject();
        project.Write(
            "applications/main/data/Repository.cs",
            "namespace Applications.Main.Data; internal sealed class Repository { }"
        );
        project.Write(
            "applications/main/ui/View.cs",
            "using Applications.Main.Data; namespace Applications.Main.Ui; internal sealed class View { internal Repository? Repository { get; } }"
        );

        var findings = Scanner.ScanPath(
            new ScanPathInput(project.PathFor("applications/main/ui/View.cs"), [])
        );

        Assert.Contains(findings, item => item.Code == "UPD101");
    }

    [Fact]
    public void SingleFileScanPreservesApplicationBoundary()
    {
        using var project = new TempProject();
        project.Write(
            "applications/settings/process/SettingsProcessing.cs",
            "namespace Applications.Settings.Process; internal static class SettingsProcessing { internal static void Run() { } }"
        );
        project.Write(
            "applications/main/process/MainCommander.cs",
            "using Applications.Settings.Process; namespace Applications.Main.Process; internal sealed class MainCommander { internal void Run() => SettingsProcessing.Run(); }"
        );

        var findings = Scanner.ScanPath(
            new ScanPathInput(project.PathFor("applications/main/process/MainCommander.cs"), [])
        );

        Assert.Contains(findings, item => item.Code == "UPD102");
    }
}
